compiler/modules/front/tests/testParser.bal

```diff
@@ -42,6 +42,63 @@ function testParser(string k, string rule, string subject, string expected) retu
     test:assertEquals(actual, expected, "wrong ast");
 }
 
+@test:Config {
+    dataProvider: sourceFragments
+}
+function testTokenizer(string k, string src) returns error? {
+    int[] lines = findLineFeeds(src);
+    Tokenizer tok = new (src);
+
+    err:Syntax|Token? t = advance(tok, k, src);
+    while t is Token {
+        err:Position pos = tok.currentPos();
+        string tStr = tokenToString(t);
+        // XXX change below after making indexInLine 1-indexed
+        int tStart = lines[pos.lineNumber - 1] + pos.indexInLine;
+        string srcAtPos = src.substring(tStart, tStart + tStr.length());
+        test:assertEquals(srcAtPos, tStr);
+        t = advance(tok, k, src);
+    }
+    if k.includes("E") {
+        test:assertTrue(t is err:Syntax, "an error expected for '" + src + "'");
+    }
+}
+
+function advance(Tokenizer tok, string k, string subject) returns err:Syntax|Token? {
+    err:Syntax? e = tok.advance();
+    if e is err:Syntax {
+        if k.includes("E") {
+            return e;
+        }
+        else {
+            panic error("tokenizer error for '" + subject + "'", e);
+        }
+    }
+    return tok.current();
+}
+
+function tokenToString(Token t) returns string {
+    if t is VariableLengthToken {
+        return t[1];
+    }
+    return <string>t;
+}
+
+function findLineFeeds(string str) returns int[] {
```

Comment:
To deal with line terminators of different lengths (2 and 1), I suggest you compute the index of the first character of every line, i.e. of the character following all the characters (2 or 1) of the line terminator.